### ISSUE TITLE
wreck job is linked in lwj-complete after failure

### ISSUE THREAD

**Issue by lipari** [flux-framework/flux-core#1101]
t2000-wreck.t 49 fails after latest merge:

expecting success: 
    test_must_fail flux wreckrun --input=bad.file hostname &&
    check_complete_link $(last_job_id)

Error: input: bad.file: No such file or directory
wreckrun: job 48 failed
2017-07-07T21:08:26.544833Z lwj.48.err[0]: Error in initialization, terminating job
flux-kvs: lwj-complete.3.48.state: No such file or directory
flux-kvs: lwj-complete.3.48.state: No such file or directory
flux-kvs: lwj-complete.3.48.state: No such file or directory
flux-kvs: lwj-complete.3.48.state: No such file or directory
flux-kvs: lwj-complete.3.48.state: No such file or directory
flux-kvs: lwj-complete.3.48.state: No such file or directory
not ok 49 - wreck job is linked in lwj-complete after failure

**Comment by grondo** [comment]
I was able to reproduce this on my desktop. My current theory is that it is a problem with the test, not with any of the code changes.

**Comment by garlick** [comment]
Hmm, that's strange, I didn't see that in any of my testing (on c9.io), and I ran the tests between each commit of the last PR...

**Comment by grondo** [comment]
I'm noticing it is intermittent and could be timing-related. I added some debug to the tests and now it is only reproducing once every 10 times... I'm still not sure what is going on though.

**Comment by grondo** [comment]
The check for "job is linked in lwj-complete" is a bit kludgy. Since the job links are archived by epoch, the script has to list the `lwj-complete.` dir and assume the job of interest is in the last epoch dir. If the order of the directories isn't sorted, or the epoch directory doesn't show up before the check, then the test will fail.

**Comment by grondo** [comment]
Yeah, I think the problem is that the check for the correct epoch dir in `lwj-complete` is only done on the first pass, then the function checks up to 5 times for the `lwj-complete.{epoch}.{id}` dir. The check for the last epoch dir should be done before each loop, not only on the first iteration. 

I don't have a clear idea why this would only start occurring now, since this check has always been 
broken. Perhaps just some sort of timing change.

@lipari -- if you are able,  you could give this small patch a whirl:

```diff
diff --git a/t/t2000-wreck.t b/t/t2000-wreck.t
index d0df918..cc9e94f 100755
--- a/t/t2000-wreck.t
+++ b/t/t2000-wreck.t
@@ -372,8 +372,8 @@ test_expect_success NO_SCHED 'flux-submit: returns ENOSYS when sched not loaded'
 '
 
 check_complete_link() {
-    lastdir=$(flux kvs dir lwj-complete | tail -1)
     for i in `seq 0 5`; do
+        lastdir=$(flux kvs dir lwj-complete | tail -1)
         flux kvs get ${lastdir}${1}.state && return 0
         sleep 0.2
     done
```

**Comment by lipari** [comment]
> @lipari -- if you are able, you could give this small patch a whirl:

Yup:
ok 49 - wreck job is linked in lwj-complete after failure
10 times in a row!  Nice diagnostic work, @grondo!

**Comment by lipari** [comment]
Closed by #1102

**Comment by chu11** [comment]
> assume the job of interest is in the last epoch dir

> I don't have a clear idea why this would only start occurring now, since this check has always been
broken. Perhaps just some sort of timing change.

I saw this error off and on while I was developing #1108.  I think there may be a reason why it's suddenly occurring (and why I may be noticing it more often in my branch).

Internally within jansson json objects, it hashes the keys/values in some manner.  When one iterates through elements in the object, you are somewhat at the will of the hashing algorithm on the order of how the objects are iterated through.  I'm guessing if this test uses ```tail -1```, there is some random luck on directory output order (which ```flux-kvs dir``` internally is using a jansson iterator).  In some newer versions of jansson, apparently insertion order is preserved, but it doesn't appear to be supported in 2.6.

I'm not sure if somewhere in flux (like in ```flux-kvs dir``` or similar) if things should be output in some sorted way?  Or is it simply bad form for this test to use ```tail -1```??  Should the lwj output be sorted in some way (via ```sort``` or otherwise) before it is tailed?

**Comment by grondo** [comment]
> I'm not sure if somewhere in flux (like in flux-kvs dir or similar) if things should be output in some sorted way? Or is it simply bad form for this test to use tail -1?? Should the lwj output be sorted in some way (via sort or otherwise) before it is tailed?

Sorting output in the test before `tail -1` can't hurt, but the problem in this issue was that the epoch directory wasn't showing up on the first iteration through the loop, and the test wasn't fetching the new list of directories after sleeping and retrying, and that has sine been fixed.

If you are seeing new failures in this test, let's open a new issue. You're correct that sorting the output of `flux kvs dir` will probably make the test more robust, but keep in mind you'll want to sort on the epoch component, not the jobid (final component) of the `.`-separated path.

**Comment by chu11** [comment]
@grondo Ahhh, I see what you mean now about the getting the epoch dir.  I'll open a new issue if I can convince myself it's not due to anything I did.

**Comment by chu11** [comment]
Re-opening this, just sent PR #1109.

What I think the prior fix did is simply increase the odds of success.  When the ```lastdir=$(flux kvs dir lwj-complete | tail -1)``` was outside of the loop, there was only a one time chance of getting the right epoch dir.  If the wrong directory was grabbed, that was it.

When placing it inside the loop, we increase the odds of success but don't guarantee it.  Later calls to ```flux kvs dir``` may output the directories in a different order, so there is an increased chance of success b/c the correct/last epoch dir was grabbed.  But we only do this 5 times, there's no guarantee of success.
